ProjFS.Mac/PrjFSKext/VirtualizationRoots.cpp

```diff
@@ -199,10 +212,15 @@ static VirtualizationRootHandle FindOrDetectRootAtVnode(vnode_t _Nonnull vnode,
     uint32_t vid = vnode_vid(vnode);
     
     VirtualizationRootHandle rootIndex;
+    bool rootVnodeStale = false;
     
     RWLock_AcquireShared(s_virtualizationRootsLock);
     {
         rootIndex = FindRootAtVnode_Locked(vnode, vid, vnodeFsidInode);
+        rootVnodeStale =
```

Comment:
Should we (or do we) need to make this check every time we call `FindRootAtVnode_Locked`? I saw we're not on line 246.

Reply:
We are holding an iocount on the vnode; so if `FindRootAtVnode_Locked()` on line 246 finds an entry, that means another thread has inserted the same vnode as a root between us dropping the shared lock on line 225 and reacquiring the exclusive lock on 243. Due to the iocount, the vnode cannot possibly have been recycled and gone stale in that time.

Reply:
Incidentally, it's no big deal if the vnode does go stale and we *don't* refresh it. I'm contemplating removing the lookup by vnode altogether, and simply always using the fsid/inode pair to identify a root. In theory, with the vnode based lookup, we can avoid fetching the inode, but the way the code is currently structured, we always fetch the inode anyway. I've not done anything radical like that for now though as the vnode cache will cause major changes in this area again anyway.